flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkBusyJoinReorderRule.java

```diff
@@ -0,0 +1,688 @@
+            newCondition.add((RexCall) resultRex);
+        }
+
+        return newCondition;
+    }
+
+    private static Optional<Tuple2<Set<RexCall>, JoinRelType>> getConditionsAndJoinType(
+            Set<Integer> oneItemIds,
+            Set<Integer> otherItemIds,
+            LoptMultiJoin multiJoin,
+            boolean isOuterJoin) {
+        if (oneItemIds.size() + otherItemIds.size() < 2) {
+            return Optional.empty();
+        }
+        JoinRelType joinType = JoinRelType.INNER;
+        if (multiJoin.getMultiJoinRel().isFullOuterJoin()) {
+            assert multiJoin.getNumJoinFactors() == 2;
+            joinType = JoinRelType.FULL;
+        }
+
+        Set<RexCall> resultRexCall = new HashSet<>();
+        List<RexNode> joinConditions = new ArrayList<>();
+        if (isOuterJoin) {
+            for (int i = 0; i < multiJoin.getNumJoinFactors(); i++) {
+                joinConditions.addAll(RelOptUtil.conjunctions(multiJoin.getOuterJoinCond(i)));
+            }
+        } else {
+            joinConditions = multiJoin.getJoinFilters();
+        }
+
+        for (RexNode joinCond : joinConditions) {
+            if (joinCond instanceof RexCall) {
+                RexCall callCondition = (RexCall) joinCond;
+                ImmutableBitSet factorsRefByJoinFilter =
+                        multiJoin.getFactorsRefByJoinFilter(callCondition);
+                int oneItemNumbers = 0;
+                int otherItemNumbers = 0;
+                for (int oneItemId : oneItemIds) {
+                    if (factorsRefByJoinFilter.get(oneItemId)) {
+                        oneItemNumbers++;
+                        if (isOuterJoin && multiJoin.isNullGenerating(oneItemId)) {
+                            joinType = JoinRelType.RIGHT;
+                        }
+                    }
+                }
+                for (int otherItemId : otherItemIds) {
+                    if (factorsRefByJoinFilter.get(otherItemId)) {
+                        otherItemNumbers++;
+                        if (isOuterJoin && multiJoin.isNullGenerating(otherItemId)) {
+                            joinType = JoinRelType.LEFT;
+                        }
+                    }
+                }
+
+                if (oneItemNumbers > 0
+                        && otherItemNumbers > 0
+                        && oneItemNumbers + otherItemNumbers
+                                == factorsRefByJoinFilter.asSet().size()) {
+                    resultRexCall.add(callCondition);
+                }
+            } else {
+                return Optional.empty();
+            }
+        }
+
+        if (resultRexCall.isEmpty()) {
+            return Optional.empty();
+        } else {
+            return Optional.of(Tuple2.of(resultRexCall, joinType));
+        }
+    }
+
+    // ~ Inner Classes ----------------------------------------------------------
+    private static class JoinPlan {
+        final LinkedHashSet<Integer> itemIds;
```

Comment:
What does `itemId` mean ?

Reply:
> What does `itemId` mean ?

In this version, I change item -> factor, factor means one multi join input.